index.d.ts

```diff
@@ -0,0 +1,212 @@
+// Type definitions for neo4j-graphql-js
+// Project: https://github.com/neo4j-graphql/neo4j-graphql-js
+// Definitions by: Nopzen <https://github.com/nopzen>
+// Definitions: https://github.com/neo4j-graphql/neo4j-graphql-js.git
+// TypeScript Version: 3.8
+
+/* eslint no-unused-vars: "off" */
+
+declare module 'neo4j-graphql-js' {
+    import { Driver } from 'neo4j-driver';
+    import { GraphQLSchema, GraphQLFieldResolver, GraphQLResolveInfo, ExecutionResult, DocumentNode } from 'graphql';
+    import { IResolvers } from 'graphql-tools';
+
+    /**
+     * makeAugmentedSchema
+     * @description Wraps {@link https://www.apollographql.com/docs/apollo-server/api/apollo-server/#makeExecutableSchema|makeExecutableSchema} to create a GraphQL schema from GraphQL type definitions (SDL). Will generate Query and Mutation types for the provided type definitions and attach `neo4jgraphql` as the resolver for these queries and mutations. Either a schema or typeDefs must be provided. `resolvers` can optionally be implemented to override any of the generated Query/Mutation fields. Additional options are passed through to `makeExecutableSchema`.
+     *
+     * @param {makeAugmentedSchemaOptions} options
+     */
+    export function makeAugmentedSchema(options: makeAugmentedSchemaOptions): GraphQLSchema;
+
+    /**
+     * neo4jgraphql
+     * @description This function's signature matches that of {@link https://graphql.org/learn/execution/#root-fields-resolvers|GraphQL resolver functions}. and thus the parameters match the parameters passed into resolve by GraphQL implementations like graphql-js.
+     *
+     * It can be called within a resolver to generate a Cypher query and handle the database call to Neo4j to completely resolve the GraphQL request. Alternatively, use `cypherQuery` or `cypherMutation` within a resolver to only generate the Cypher query and handle the database call yourself.
+     * @param {object} object                   The previous object being resolved. Rarely used for a field on the root Query type.
+     * @param {RequestArguments} args           The arguments provided to the field in the GraphQL query.
+     * @param {Neo4jContext} context            Value provided to every resolver and hold contextual information about the request, such as the currently logged in user, or access to a database. neo4j-graphql-js assumes a neo4j-javascript-driver instance exists in this object, under the key driver.
+     * @param {GraphQLResolveInfo} resolveInfo  Holds field-specific information relevant to the current query as well as the GraphQL schema.
+     * @param {boolean} debug                   Specifies whether to log the generated Cypher queries for each GraphQL request. Logging is enabled by default.
+     */
+    export function neo4jgraphql(
+        object: any,
+        args: RequestArguments,
+        context: Neo4jContext,
+        resolveInfo: GraphQLResolveInfo,
+        debug?: boolean,
+    ): ExecutionResult;
```

Comment:
I think the result here is `Promise<ExecutionResult>` and not `ExecutionResult`

Reply:
Also, couldn't we get the result properly typed by using a generic ?

```ts
    export function neo4jgraphql(
        object: T,
        args: RequestArguments,
        context: Neo4jContext,
        resolveInfo: GraphQLResolveInfo,
        debug?: boolean,
    ): Promise<ExecutionResult<T>>;
```

Reply:
Actually, I'm realizing that neo4jgraphql does NOT return a ExecutionResult.
When I console.log the result with a `await no4jgraphql`, I see
```json
RESULT {
  FRAGMENT_TYPE: 'FlatPage',
  id: 'settings',
  title: 'Settings page',
  slug: 'settings'
}
```
Which is the type `TData & { FRAGMENT_TYPE: string }`

This makes me believe that a correct typing of the function (also typing the result properly with a generic) would be:
```ts
  export function neo4jgraphql<TData>(
    object: TData,
    args: RequestArguments,
    context: Neo4jContext,
    resolveInfo: GraphQLResolveInfo,
    debug?: boolean
  ): Promise<TData & { FRAGMENT_TYPE: string }>;
```

To call with
```ts
      const result = await neo4jgraphql<IPage>(page, params, context, resolveInfo);
```

Reply:
@johnymontana can you confirm this ? Because [the doc](https://grandstack.io/docs/neo4j-graphql-js-api/#neo4jgraphqlobject-params-context-resolveinfo-debug-executionresult) incorrectly states that neo4jgraphql returns an ExecutionResult, which is typed as:
```ts
/**
 * The result of GraphQL execution.
 *
 *   - `errors` is included when any errors occurred as a non-empty array.
 *   - `data` is the result of a successful execution of the query.
 *   - `extensions` is reserved for adding non-standard properties.
 */
export interface ExecutionResult<
  TData = { [key: string]: any },
  TExtensions = { [key: string]: any }
> {
  errors?: ReadonlyArray<GraphQLError>;
  // TS_SPECIFIC: TData. Motivation: https://github.com/graphql/graphql-js/pull/2490#issuecomment-639154229
  data?: TData | null;
  extensions?: TExtensions;
}
```
There is certainly no `data`, `errors` or `extensions` key in the result from the function as of neo4j-graphql-js 2.19.2.

Reply:
These types was created over 1 year ago :D maybe this have changed, It is not until recently I picked this back up once @johnymontana talked about getting this in to the codebase.

But I've you have seen the signature updated, I guess I can have a look into this as well today after work :)

Reply:
Well.. it's been incorrect for about a year as well.. just found back #447 .

Thanks for looking into it <3

Reply:
So sorry @AdrienLemaire, have had a busy last month, currently not at my laptop to fix this gain before around the 16 this month, so another week or so.

Reply:
@AdrienLemaire Please see latest commit 7245462 I believe that should address your comments? if this is correct please resolve this comment :)

sorry for the day, was on a extended vacation without my laptop with this code on :(

Reply:
I don't have the authorization to resolve comments, but LGTM

Reply:
Cool, I asked @johnymontana  if he could have one of the Neo4j guys looking at it. I will try and keep my fork upto date so people who would like to use this types can use my fork of the repo till it gets merged.

Reply:
> Cool, I asked @johnymontana if he could have one of the Neo4j guys looking at it. I will try and keep my fork upto date so people who would like to use this types can use my fork of the repo till it gets merged.

https://github.com/neo4j-graphql/neo4j-graphql-js/issues/608